file_formats/cmt_format.ml

```diff
@@ -174,10 +322,12 @@ let save_cmt filename modname binary_annots sourcefile initial_env cmi shape =
            | None -> None
            | Some cmi -> Some (output_cmi temp_file_name oc cmi)
          in
+         gather_declarations binary_annots;
+         let cmt_annots = clear_env binary_annots in
```

Comment:
This looks like you are going to lose sharing between the `cmt_annots` and the declarations in the table. Why not remove the environments first and then call `gather_declarations`? That way you can remove `clear_decl` and keep the sharing.

Reply:
We need the shapes that are in the environment to reduce shapes when indexing. Having intermediate shapes stripped off the environment in cmt files is what led to #11983 and to this PR.

We could, however, do it in two iterations:
- first iterate for the index
- then clean the env before iterating for the `uid_to_decl` table

(I am currently trying to reproduce a case where we "need the shapes for the intermediate modules")

Reply:
I tried doing it in two iterations: this doesn't change the build time nor the size of the artifacts.
I think sharing is still working as expected, independently of the order. In the previous PR we saw a cmt size increase because I was storing also parts of the Types tree instead of only Typedtree nodes at the beginning.

Reply:
I find that very surprising. Unless I'm missing something, it seems slightly more likely that you were still losing sharing on the new version than somehow keeping it on this version.

Reply:
For reference, the "ensure sharing" version I used is: https://github.com/voodoos/ocaml/commit/5cc4b2e7c2dcaaa5875733b7ae9f6dcf9414544c

But I realize now that I still re-clean the environment in this version, I will stop doing that and try again.